### DISCUSSION TITLE
Is there a way to validate .pest grammars?

### DISCUSSION THREAD

**Discussion by matthew-dean** [pest-parser/pest#617] [General]
I've worked a bit with other parser generators (parsers that operate from a PEG-style grammar), and typically there's a CLI tool that can catch / point out errors in the grammar. Is the only way to test a grammar to just try it and see if Rust compilation breaks? I notice that even the editor at pest.rs will throw unhelpful "unreachable" errors if something is wrong.

**Comment by RXminuS** [comment]
I think this is definitely one of the shortcomings right now. We basically develop our grammars on https://pest.rs/#editor which has led to some big frustrations when we accidentally navigate away from the page and all our hard work has vanished.
I have a half-arsed project somewhere to make a better stand-alone online editor (like regex-101) but there were too many moving bits with wasm etc from the landing page that I gave up on it for now as I simply didn't have the time...but definitely think this is something we should address sooner rather than later and would love to sponsor if someone feels like picking it up.

**Comment by activexray** [comment]
I'm using [pest-mode](https://github.com/ksqsf/pest-mode) in emacs and it works beautifully. Having a test repl with the ast next to it is brilliant. I can't imagine trying to iterate on grammar without it.
